File: exercises/D107/solution.java
import java.util.*;

class D107 {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        long n = in.nextLong();
        long a = Math.round(Math.sqrt(n));
        long b = (long)Math.floor(Math.sqrt(n * 2));
        boolean square = a * a == n;
        boolean triangular = b * (b + 1) / 2 == n;
        if (square && triangular) {
            System.out.println("Both");
        } else if (square) {
            System.out.println("Square");
        } else if (triangular) {
            System.out.println("Triangular");
        } else {
            System.out.println("Neither");
        }
    }
}

File: exercises/D107/solution.cpp
#include <cmath>
#include <iostream>
using namespace std;
int main() {
    long long n;
    cin >> n;
    long long a = round(sqrt(n));
    long long b = floor(sqrt(n * 2));
    bool square = a * a == n;
    bool triangular = b * (b + 1) / 2 == n;
    if (square && triangular) {
        cout << "Both" << endl;
    } else if (square) {
        cout << "Square" << endl;
    } else if (triangular) {
        cout << "Triangular" << endl;
    } else {
        cout << "Neither" << endl;
    }
}
